### DISCUSSION TITLE
INSERT INTO table errors out when using MappedAsDataclass (There is an entry for table "user", but it cannot be referenced from this part of the query)

### DISCUSSION THREAD

**Discussion by MRogelio** [sqlalchemy/sqlalchemy#11497] [Usage Questions] [answered]
I'm using sqlalchemy 2.0.30, FastAPI 0.111.0, Pydantic 2.7.1 and `postgresql+asyncpg` for the engine.
When I use `MappedAsDataclass` on my model with `Column` I get the following error:
```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedTableError'>: invalid reference to FROM-clause entry for table "user"
DETAIL:  There is an entry for table "user", but it cannot be referenced from this part of the query.
[SQL: INSERT INTO "user" (user_id, user_name) VALUES ("user".user_id, $1::VARCHAR) RETURNING "user".user_id]
[parameters: ('string',)]
```

If I use `mapped_column` instead of `Column` the issue is that it requires a `user_id` for the `UserModel` I insert in the database although it's specified as `autoincrement` and it's not possible to give it because the column is `GENERATED ALWAYS AS IDENTITY`.
If I use `Colomn` and remove the `MappedAsDataclass` it works but that's not an ideal solution as it's convenient in other parts of my code.
If I use `mapped_colomn`, remove `GENERATED ALWAYS AS IDENTITY` and give it as an argument when I create a `UserModel` instance it works but this solution is less than ideal as I lose all the benefit of having an `autoincrement` colomn.
I looked up this issue and only found something similar [here](https://github.com/sqlalchemy/sqlalchemy/issues/8718) but it was apparently a bug and it was fixed. Am I doing something wrong here?
Here is the code I end up with when I tried to isolate the issue with the bare minimum.

SQL table
```
CREATE TABLE "user" (
    "user_id" INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    "user_name" VARCHAR(64) NOT NULL)
);
```

model
```
class UserModel(MappedAsDataclass, CustomBase):
    __tablename__ = "user"
    user_id: Mapped = Column(Integer, primary_key=True, autoincrement=True)
    user_name: Mapped = Column(String(64), nullable=False)
```

schema
```
class UserCreate(BaseModel):
    user_name: str
    
 class UserSchema(UserCreate):
    user_id: int
```

endpoint
```
@router.post(
    "/user",
    tags=["User"],
    description="Create a user",
    response_model=UserCreate,
)
async def user_add(requ: UserCreate) -> UserCreate:
    session = async_session_factory()
    async with session:
        new_user = UserModel(**requ.model_dump())
        session.add(new_user)
        await session.commit()
    return new_user
```

**Comment by zzzeek** [comment]
I dont know what `model_dump()` does but it looks like it's saying something like `user_id=User.user_id`, that is, assignment to an expression, which is wrong.

I dont really know how it's doing that because I dont know anything about what `model_dump()` produces, but the bug would be there in some way.  so you have some kind of pydantic issue to check on.

edit: also your use of mappedasdataclass is incorrect, dont use Column

**Reply by zzzeek** [reply, marked as answer]
here's a correct program, run this

```py
from __future__ import annotations

import asyncio

from sqlalchemy import String
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import MappedAsDataclass


class Base(DeclarativeBase):
    pass


class UserModel(MappedAsDataclass, Base):
    __tablename__ = "user"
    user_id: Mapped[int] = mapped_column(
        primary_key=True, autoincrement=True, init=False
    )
    user_name: Mapped[str] = mapped_column(String(64), nullable=False)


e = create_async_engine(
    "postgresql+asyncpg://scott:tiger@localhost/test", echo=True
)


async def main():
    async with e.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    await user_add()


async def user_add():
    async with AsyncSession(e) as session:
        new_user = UserModel(user_name="name")
        session.add(new_user)
        await session.commit()
    return new_user

asyncio.run(main())

```
